### ISSUE TITLE
ConvexHull example is broken

### ISSUE THREAD

**Issue by WestLangley** [mrdoob/three.js#10624]
`ConvexGeometry` appears to be generating extra internal faces.

There are calls to `Math.random()` in the algorithm, so the output is not deterministic. Several refreshes may be required to see the issue.

![screen shot 2017-01-21 at 4 07 37 pm](https://cloud.githubusercontent.com/assets/1000017/22177846/9443983c-dff4-11e6-98c0-e17d0300c5ab.png)

Only tested on macOS.

##### Three.js version

- [ ] Dev
- [ ] r82 and later
- [ ] ...

/ping @Mugen87 Maybe this problem would interest you? ;-)

**Comment by Mugen87** [comment]
I'll put this on my todo list 😉 

Maybe we can directly port this to `BufferGeometry` when fixing the bug...

**Comment by Mugen87** [comment]
The bug actually disappears if i delete the following [lines](https://github.com/mrdoob/three.js/blob/dev/examples/js/geometries/ConvexGeometry.js#L37-L40). I'm not 100% why this offsetting is performed. @WestLangley Do you have any ideas?

Anyway, i think a typical implementation of the incremental convex hull algorithm looks a little bit different than in `ConvexHull`. Randomly shifting vertices should not be necessary. Besides, the actual time complexity should be O(nlogn) and not O(n^2).

For example the visibility test is normally done by calculating the volume of a tetrahedron that consist of three face vertices and the current vertex to add. The sign of the volume indicates if the vertex is before or behind the face. If you perform this test for all existing faces, you can calculate the horizon and the shape of the next intermediate hull.

Implementing this will take some time. Maybe it's better to provide a quick fix first....

**Comment by WestLangley** [comment]
>The bug actually disappears if i delete the following lines.

What I don't understand, is the bug does not appear to occur before rev 82. At least, I don't think it does...

>Implementing this will take some time. Maybe it's better to provide a quick fix first....

Maybe a replacement would be better. See this thread: https://github.com/mrdoob/three.js/pull/1841.

/ping @marklundin

**Comment by donmccurdy** [comment]
@Mugen87 the random offsetting is common in convex hull calculation, to deal with edge cases around co-planar points, or perhaps duplicate points. I haven't implemented this incremental algorithm before (just a 2D MergeHull once) but you can reproduce the problem easily enough with a box. This snippet works only with that randomization:


```
geometry = new THREE.ConvexGeometry(new THREE.BoxGeometry(5, 5, 5).vertices);
```

**Comment by Mugen87** [comment]
@donmccurdy Thanks for the explanation. I've actually never seen the usage of `Math.random()` in an convex hull algorithm before.

Can you say something about the numerical stability and robustness of the linked Quickhull implementation? Some time ago, i've tried to develop a Quickhull 3D port for `three.js` based on this [presentation](http://media.steampowered.com/apps/valve/2014/DirkGregorius_ImplementingQuickHull.pdf) of Valve. But i'v never finished this piece of code, because it was quite sophisticated to regard all possible invariants (see page 74).

BTW: The presentation is excellent! 😊

**Comment by WestLangley** [comment]
Another JS implementation: https://github.com/maurizzzio/quickhull3d

with three.js r.73 demo: http://requirebin.com/?gist=9b19fccfa670c9e2597b

**Comment by WestLangley** [comment]
Fixed in #10987
